src/Microsoft.Azure.WebJobs.Host/Bindings/Path/BindingTemplateToken.cs

```diff
@@ -1,32 +1,271 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Globalization;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CSharp.RuntimeBinder;
+using Newtonsoft.Json.Linq;
+
 namespace Microsoft.Azure.WebJobs.Host.Bindings.Path
 {
     /// <summary>
     /// Represents a semantic token as generated by <see cref="BindingTemplateParser"/> during parsing of parameterized 
     /// binding template string. Then generated token may be consumed to create instance of 
     /// <see cref="BindingTemplate"/> or <see cref="BindingTemplateSource"/>.
+    /// The token may be a literal value or an expression. 
     /// </summary>
-    internal struct BindingTemplateToken
+    internal abstract class BindingTemplateToken
     {
-        private readonly string _value;
-        private readonly bool _isParameter;
+        // If this is expression, get the prameter name 
+        // If it's a literal, return null.
+        public abstract string ParameterName { get; }
+
+        // If this is a literal, return the value. Else, return null.
+        public abstract string AsLiteral { get; }
+
+        public static BindingTemplateToken NewLiteral(string literalValue)
+        {
+            return new LiteralToken(literalValue);
+        }
+
+        public static BindingTemplateToken NewExpression(string expression)
+        {
+            if (!BindingTemplateParser.IsValidIdentifier(expression))
```

Comment:
I think we should rename IsValidIdentifier to ValidateTemplate, since we've been expanding the syntax over time